django/db/models/query.py

```diff
@@ -2450,6 +2450,12 @@ def has_to_attr_attribute(instance):
             # singly related object, descriptor object has the
             # get_prefetch_queryset() method.
             if hasattr(rel_obj_descriptor, "get_prefetch_queryset"):
+                # The name is used to define the cache. Setting it to the
+                # to_attr makes the caching work properly for singly
+                # related objects
+                if hasattr(rel_obj_descriptor, "field"):
+                    rel_obj_descriptor.field.name = to_attr
```

Comment:
@charettes I changed it to simply keep the default `is_fetched` in the case where `through_attr` is not the same as `to_attr`. I believe that should only happen in the case where a `to_attr` parameter was set, so that might be enough. Let me know if there is something I did not think about on this.

Note that this does mean that something like 
`prefetch_related(
    'pizza__toppings',
    Prefetch(
        'pizza__toppings',
        to_attr='something_else',
    )`
would cause 2 calls to the database, because it would not use the cache. I think this is still preferable to `something_else` not existing at all. This is also a very weird use case for which I see no real use ever, so I think it should be fine.

Reply:
@MaxDude132 I also agree that this is an acceptable compromise and if the test suite agrees it's a good sign that it was never an expected behaviour 👍